dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,263 @@
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING:
+        case WNP_E_BIND_USER_BUSY:
+        case RPC_S_SERVER_UNAVAILABLE:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    inline bool IsPackagedProcess()
+    {
+        wchar_t packageFullName[PACKAGE_FULL_NAME_MAX_LENGTH + 1] = {};
+
+        UINT32 packageFullNameLength = static_cast<UINT32>(ARRAYSIZE(packageFullName));
+
+        const auto packagedProcessHResult = ::GetCurrentPackageFullName(&packageFullNameLength, packageFullName);
+
+        if (packagedProcessHResult == APPMODEL_ERROR_NO_PACKAGE)
+        {
+            return false;
+        }
+
+        THROW_IF_WIN32_ERROR(packagedProcessHResult);
+
+        return true;
+    }
+
+    winrt::IAsyncOperationWithProgress<winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelResult, winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelStatus> PushNotificationManager::CreateChannelAsync(winrt::guid remoteId)
+    {
+        static bool s_remoteIdInProgress;
+        static wil::critical_section s_lock;
+
+        THROW_HR_IF(E_INVALIDARG, (remoteId == winrt::guid()));
+
+        // API supports channel requests only for packaged applications for v0.8 version
+        THROW_HR_IF(E_NOTIMPL, (IsPackagedProcess() == false));
+
+        winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelResult channelResult{ nullptr };
+        {
+            auto lock = s_lock.lock();
+            if (s_remoteIdInProgress == false)
+            {
+                s_remoteIdInProgress = true;
+            }
+            else
+            {
+                channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                    nullptr, WPN_E_OUTSTANDING_CHANNEL_REQUEST, PushNotificationChannelStatus::CompletedFailure);
+                co_return channelResult;
+            }
+        }
+
+        auto scopeExit = wil::scope_exit([&]()
+            {
+                auto lock = s_lock.lock();
+                s_remoteIdInProgress = false;
+            });
+
+        auto cancellation{ co_await winrt::get_cancellation_token() };
+
+        cancellation.enable_propagation(true);
+
+        // Allow to register the progress and complete handler
+        co_await resume_background();
+
+        auto progress{ co_await winrt::get_progress_token() };
+
+        uint8_t retryCount = 0;
+        winrt::hresult channelRequestResult = E_PENDING;
+        PushNotificationChannelStatus status = PushNotificationChannelStatus::InProgress;
+
+        PushNotificationCreateChannelStatus
+            channelStatus = { channelRequestResult, status, retryCount };
+
+        progress(channelStatus);
+
+        PushNotificationChannelManager channelManager{};
+        winrt::PushNotificationChannel pushChannelReceived{ nullptr };
```

Comment:
No need to predeclare this variable. It is used in only one place - declare at point of use.